src/data_readers/data_reader_hdf5.cpp

```diff
@@ -0,0 +1,227 @@
+    hsize_t dims_local[4] = {xPerNode, yPerNode, zPerNode, cPerNode};
+
+    // necessary for the hdf5 lib
+    hid_t memspace = H5Screate_simple(4, dims_local, NULL);
+    int odd_offset;
+    int spatial_offset = rank%num_io_parts
+    if (xlines > 1) {
+      // this is theoretically the odd case but it has not been tested yet
+      // so everything with odd should be ignored
+      if(spatial_offset < int(dims[0]%xPerNode)) {
+        offset[0] = ((xPerNode+1)*spatial_offset);
+      } else {
+        // offset of this x dim for this rank;
+        // in most cases odd_offset will be 0
+        odd_offset = (xPerNode+1)*(dims[0]%xPerNode);
+        offset[0] = odd_offset +(xPerNode*(spatial_offset%xlines)); 
+      }
+    } else {
+      offset[0] = 0;
+    }
+
+    if (ylines > 1) {
+      if(spatial_offset < int(dims[1]%yPerNode)) {
+        offset[1] = ((yPerNode+1)*spatial_offset);
+      } else {
+        // offset of the y dim for this rank
+        odd_offset = (yPerNode+1)*(dims[1]%yPerNode);
+        offset[1] = odd_offset + (yPerNode*(spatial_offset/ylines));
+      }   
+    } else {
+      offset[1] = 0;
+    }
+
+    if (zlines > 1) {
+      offset[2] = zPerNode*spatial_offset;
+    } else {
+      offset[2] = 0;
+    }
+    // I have channel dim splits 
+    // all combos arent really tested
+    //add the rest later
+    offset[3] = 0; 
+    // from an explanation of the hdf5 select_hyperslab:
+    // start -> a starting location for the hyperslab
+    // stride -> the number of elements to separate each element or block to be selected
+    // count -> the number of elemenets or blocks to select along each dimension
+    // block -> the size of the block selected from the dataspace 
+    //hsize_t status;
+
+    //todo add error checking
+    H5Sselect_hyperslab(filespace, H5S_SELECT_SET, offset, NULL, count, dims_local);
+    H5Dread(h_data, H5T_NATIVE_SHORT, memspace, filespace, H5P_DEFAULT, data_out);
+  }
+
+  void hdf5_reader::load() {
+    lbann_comm* l_comm = get_comm();
+    const El::mpi::Comm & w_comm = l_comm->get_world_comm();
+    MPI_Comm mpi_comm = w_comm.GetMPIComm();
+    int world_rank = get_rank_in_world();
+    int color = world_rank/dc::get_number_of_io_partitions(); 
+    MPI_Comm_split(mpi_comm, color, world_rank, &m_comm);
+    m_shuffled_indices.clear();
+    m_shuffled_indices.resize(m_file_paths.size());
+    std::iota(m_shuffled_indices.begin(), m_shuffled_indices.end(), 0);
+
+    select_subset_of_data();
+  }
+  bool hdf5_reader::fetch_label(Mat& Y, int data_id, int mb_idx) {
+    return true;
+  }
+  bool hdf5_reader::fetch_datum(Mat& X, int data_id, int mb_idx) {
+    prof_region_begin("fetch_datum", prof_colors[0], false);
+    int nprocs;
+    MPI_Comm_size(MPI_COMM_WORLD, &nprocs); 
+    if ((nprocs%dc::get_number_of_io_partitions()) !=0) {
+      std::cerr<<"nprocs should be divisible by num of io partitions otherwise this wont work \n";
+    }
```

Comment:
Let's not do this every time a datum is read. It can be just done once, perhaps in `load()`.